doc/sue.1
.\" $Id: sue.1,v 1.6 2024/07/12 22:56:16 tom Exp $
.TH SUE 1 2024-07-12 MiscTools "User commands"
.ie \n(.g \{\
.ds `` \(lq
.ds '' \(rq
.\}
.el \{\
.ie t .ds `` ``
.el   .ds `` ""
.ie t .ds '' ''
.el   .ds '' ""
.\}
.hy 0
.SH NAME
sue \-
su with environment
.SH SYNOPSIS
.B sue
.I command
.
.SH DESCRIPTION
Provides a \*(``su\*('' command which preserves caller's environment.
.PP
\fBSue\fR may be a wrapper for \fBsudo\fP (e.g., to the \fIroot\fP user).
For non-root use, the setuid configuration is preferred,
because \fBsudo\fP makes it cumbersome to reuse the current \fBPATH\fP.
.
.SH OPTIONS
None.
.
.
.SH BUGS
.
Probably.
.
.SH EXAMPLES
.
\fBSue\fR is installed by convention as the user to which it should \*(``su\*(''
.
.SH SEE ALSO
sudo, su


CHANGES
2025-09-11  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/sue/sue.c: workaround for broken NetBSD header

	* td_getopt.h: NetBSD lint

	* package/debian/changelog, package/misc_tools.spec: bump

2025-09-10  Thomas E. Dickey  <dickey@invisible-island.net>

	* COPYING.asc: sign

	* package/misc_tools.spec: SPDX equivalent

	* COPYING: consolidate license

	* install-bin.in:
	amend $INSTALL in case install-sh is used from a subdirectory

	* install-bin.in: shellcheck warnings

	* config_h.in, doc/minstall.in, package/debian/postinst,
	  src/newpath/makefile.wnt, configure.in:
	trim-blanks

	* src/unmap/chrcount.c, src/unmap/map.c, src/unmap/map_s.c,
	  src/unmap/unmap.c, src/unmap/unmap_s.c, src/width/width.c,
	  src/splitit/splitit.c, src/perror/perror.c, src/slowcat/slowcat.c,
	  src/sue/sue.c, src/tisplit/timerge.c, src/tisplit/tisplit.c,
	  src/acsplit/acmerge.c, src/count_files/count_files.c, src/cpd/cpd.c,
	  src/newest/newest.c, src/newpath/newpath.c, src/acsplit/acsplit.c,
	  td_getline.h:
	s/0/NULL/

	* doc/newpath.1: bump

	* doc/newpath.1: grammar

	* doc/cpd.1: bump

	* doc/cpd.1: grammar

	* doc/chrcount.1: bump

	* doc/chrcount.1: grammar

	* doc/acsplit.1: bump

	* doc/acsplit.1: grammar

	* package/debian/copyright: bump

	* configure: regen

	* package/debian/changelog, package/misc_tools.spec: bump

2025-08-05  Thomas E. Dickey  <dickey@invisible-island.net>

	* aclocal.m4: resync with my-autoconf

2025-06-08  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-sh: resync with my-autoconf

2024-07-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: update to 2024-07-27

2024-07-15  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-bin.in: source parameter may have $DESTDIR prefix

	* src/sue/use-usual:
	use $DestDir rather than $DESTDIR, and use the install program
	to install the prepared wrapper rather than fixing after install.

	* package/misc_tools.spec: use --with-execdir option

	* src/sue/use-hacks: quoting per shellcheck warnings
	do not exit on the first error, to work with rpmbuild

	* package/debian/changelog: bump

2024-07-13  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/newpath/newpath.c:
	a relative link for newpath, i.e., making /usr/bin/newpath a symbolic link
	to ../libexec/misctools/newpath did not work -- fix this by finding the
	program on the path, and using realpath to resolve links in that.

	* install-bin.in: corrected link source

	* package/debian/postinst: using libexec now, rather than lib

	* src/sue/use-hacks:
	s/DESTDIR/DestDir/ to avoid conflict with $DESTDIR used in package builds

	* configure: regen

	* src/sue/use-hacks: exit on the first error

	* install-bin.in: add special case for src/sue

	* configure.in: add install_bin symbol, for special case in src/sue

	* src/sue/makefile.in:
	bin-directory depends upon the --with-execdir option

	* configure: regen

	* install-bin.in: account for DESTDIR

	* configure.in: improve message-result for logging

	* package/debian/postinst:
	symbolic link is now done via configure option --with-execdir

	* package/debian/rules: use --with-execdir option

	* configure: regen

	* aclocal.m4: add CF_PROG_LN_S

	* install-bin.in: RCS_BASE

	* package/misc_tools.spec:
	simplify by adding a configure option --with-execdir

	* configure.in, src/acsplit/makefile.in, src/count_files/makefile.in,
	  src/cpd/makefile.in, src/hex/makefile.in, src/isatty/makefile.in,
	  src/newest/makefile.in, src/newpath/makefile.in,
	  src/perror/makefile.in, src/realpath/makefile.in,
	  src/slowcat/makefile.in, src/splitit/makefile.in, src/sue/makefile.in,
	  src/tisplit/makefile.in, src/unmap/makefile.in, src/width/makefile.in:
	added --with-execdir option to simplify installing the tools into libexecdir
	while making a symbolic link to newpath

	* makefile.in: provide for cleanup of install-bin script

	* package/misc_tools.spec:
	build all programs, since there is no naming conflict now

	* doc/minstall.in: update to match unquoted section number in .TH line

	* package/debian/changelog, package/misc_tools.spec: bump

2024-07-12  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/misc_tools.spec:
	add relative symlink for newpath to the %files list

	* doc/newpath.1:
	explain how to reduce environment conflicts with this set of utilities

	* doc/acmerge.1, doc/acsplit.1, doc/chrcount.1, doc/count_files.1,
	  doc/cpd.1, doc/hex.1, doc/isatty.1, doc/map.1, doc/newest.1,
	  doc/perror.1, doc/realpath.1, doc/slowcat.1, doc/splitit.1, doc/sue.1,
	  doc/timerge.1, doc/tisplit.1, doc/unmap.1, doc/width.1:
	identify source (in lower-left) as "MiscTools"

	* doc/acmerge.1, doc/acsplit.1, doc/chrcount.1, doc/count_files.1,
	  doc/cpd.1, doc/hex.1, doc/isatty.1, doc/map.1, doc/newest.1,
	  doc/perror.1, doc/realpath.1, doc/slowcat.1, doc/splitit.1, doc/sue.1,
	  doc/timerge.1, doc/tisplit.1, doc/unmap.1, doc/width.1:
	quoting

	* doc/newpath.1: identify source (in lower-left) as "MiscTools"

	* doc/acmerge.1, doc/acsplit.1, doc/chrcount.1, doc/count_files.1,
	  doc/cpd.1, doc/hex.1, doc/isatty.1, doc/map.1, doc/newest.1,
	  doc/newpath.1, doc/perror.1, doc/realpath.1, doc/slowcat.1,
	  doc/splitit.1, doc/sue.1, doc/timerge.1, doc/tisplit.1, doc/unmap.1,
	  doc/width.1:
	check-manpage

	* package/debian/copyright: bump

	* configure: regen

	* COPYING.asc: sign

	* COPYING, package/debian/changelog, package/misc_tools.spec: bump

2024-06-03  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: update to 2024-05-27

2023-12-27  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: 2023/12/27

2023-12-02  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2023/12/2

2023-12-01  Thomas E. Dickey  <dickey@invisible-island.net>

	* aclocal.m4: resync with my-autoconf

2023-04-14  Thomas E. Dickey  <dickey@invisible-island.net>

	* doc/minstall.in: reduce filename-conflicts by installing ".1mt" files

	* configure: regen


src/sue/sue.c
/*
 * $Id: sue.c,v 1.7 2025/09/11 08:17:35 tom Exp $
 *
 * Author:	T.E.Dickey
 * Created:	23 Apr 1990
 * Purpose:	provide a "su" command which preserves caller's environment.
 */
#include	<stdlib.h>
#include	<unistd.h>	/* may define _POSIX_SAVED_IDS */
#include	<stdio.h>
#include	<string.h>
#include	<pwd.h>

static void
failed(const char *caller)
{
    perror(caller);
    exit(EXIT_FAILURE);
}

int
main(int argc, char *argv[])
{
    static char the_default_shell[] = "/bin/sh";
    char home[BUFSIZ];
    char shell[BUFSIZ];
    char user[BUFSIZ];
    char *default_shell = the_default_shell;
    int code = EXIT_SUCCESS;
    register struct passwd *q;

#if !defined(_POSIX_SAVED_IDS) && !defined(_NETBSD_SOURCE)
    /*
     * This works _everywhere_ except for glibc 2.1.3, which is broken.
     * Later, NetBSD also is broken.
     */
    if (setuid(geteuid()) < 0)
	failed("setuid");
    if (setgid(getegid()) < 0)
	failed("setgid");
#else
    /*
     * For the moment (2000/12/28), no one's broken the "old" behavior.
     */
    if (setreuid(geteuid(), geteuid()) < 0)
	failed("setreuid");
    if (setregid(getegid(), getegid()) < 0)
	failed("setregid");
#endif
    strcpy(home, "HOME=");
    strcpy(shell, "SHELL=");
    strcpy(user, "USER=");
    if ((q = getpwuid(getuid())) != NULL) {
	if (q->pw_shell && q->pw_shell[0])
	    default_shell = q->pw_shell;
	strcat(home, q->pw_dir);
	strcat(shell, default_shell);
	strcat(user, q->pw_name);
    } else {
	strcat(home, ".");
	strcat(shell, default_shell);
	strcat(user, getenv("LOGNAME"));
    }
    putenv(home);
    putenv(shell);
    putenv(user);
    if (argc > 1) {
	execvp(argv[1], argv + 1);
	perror(argv[1]);
    } else {
	if (system(default_shell))
	    code = EXIT_FAILURE;
    }
    exit(code);
}
